core/client/src/main/java/alluxio/client/block/RetryHandlingBlockWorkerClient.java

```diff
@@ -112,7 +126,12 @@ public void releaseClient(BlockWorkerClientService.Client client) {
   @Override
   public void close() {
     if (mHeartbeat != null) {
-      mHeartbeat.cancel(true);
+      HEARTBEAT_CANCEL_POOL.submit(new Runnable() {
+        @Override
+        public void run() {
+          mHeartbeat.cancel(true);
+        }
+      });
```

Comment:
@calvinjia actually, the deadlock is not likely to happen in the current setup because the heartbeat will yield the thread when it is not active. In this sense, I am fine with using a single threadpool though I personally prefer two threadpool to be safe.